### PR TITLE
[GPU] Added shape agnostic NonZero ref kernel

### PR DESCRIPTION
**Pull request by andrew-k-park**
### Details:
 - Shape agnostic NonZero ref kernel implementation

### Tickets:
 - 95104

### CODE REVIEW

**Review by ahnyoung-paul — APPROVED** [review]
LGTM

**Review by e-ddykim — APPROVED** [review]
Looks good to me

File: src/plugins/intel_gpu/src/graph/impls/ocl/non_zero.cpp [hunk 1]
```diff
@@ -29,11 +29,17 @@ struct count_nonzero_impl : typed_primitive_impl_ocl<count_nonzero> {
         return make_unique<count_nonzero_impl>(*this);
     }
 
+public:
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/impls/ocl/non_zero.cpp:32]: Why? Methods below are already public

File: src/plugins/intel_gpu/src/graph/impls/ocl/non_zero.cpp [hunk 2]
```diff
@@ -48,20 +54,26 @@ struct gather_nonzero_impl : typed_primitive_impl_ocl<gather_nonzero> {
         return make_unique<gather_nonzero_impl>(*this);
     }
 
+public:
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/impls/ocl/non_zero.cpp:57]: Same as above

File: src/plugins/intel_gpu/src/graph/impls/ocl/non_zero.cpp [hunk 3]
```diff
@@ -48,20 +54,26 @@ struct gather_nonzero_impl : typed_primitive_impl_ocl<gather_nonzero> {
...
     static kernel_params_t get_kernel_params(const kernel_impl_params& impl_param) {
         auto params = get_default_params<kernel_selector::gather_nonzero_params>(impl_param);
         auto optional_params = get_default_optional_params<kernel_selector::gather_nonzero_optional_params>(impl_param.get_program());
 
         params.inputs.push_back(convert_data_tensor(impl_param.get_input_layout(1)));
-        params.ov_input_rank = impl_param.get_input_layout().get_shape().size();
+        params.ov_input_rank = impl_param.get_input_layout().get_partial_shape().rank().get_length();
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/impls/ocl/non_zero.cpp:63]: `get_partial_shape().size()` can be used instead

**Review by vladimir-paramuzov — APPROVED** [review]
Overall, LGTM. Some minor comments are left

File: src/plugins/intel_gpu/src/graph/impls/ocl/non_zero.cpp (hunk 1 above)
Comment (andrew-k-park) [reply src/plugins/intel_gpu/src/graph/impls/ocl/non_zero.cpp]: removed

File: src/plugins/intel_gpu/src/graph/impls/ocl/non_zero.cpp (hunk 2 above)
Comment (andrew-k-park) [reply src/plugins/intel_gpu/src/graph/impls/ocl/non_zero.cpp]: removed

File: src/plugins/intel_gpu/src/graph/impls/ocl/non_zero.cpp (hunk 3 above)
Comment (andrew-k-park) [reply src/plugins/intel_gpu/src/graph/impls/ocl/non_zero.cpp]: updated. thanks

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/gather_nonzero_ref.cl [hunk 4]
```diff
@@ -9,18 +9,27 @@
...
+    __global OUTPUT_TYPE* output)
 {
     int local_offset = 0;
     const int result_size = OV_INPUT_RANK * OUTPUT_FEATURE_NUM; // output shape: [ov_rank, count_nonzero]
 
-#ifdef USE_LOCAL_MEM
+#if IS_DYNAMIC
+    const int dst_slm_size = TOTAL_DATA_SIZE * OV_INPUT_RANK;
+    #define DST_SLM_SIZE dst_slm_size
+#if DST_SLM_SIZE < MAX_LOCAL_MEM_SIZE
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/gather_nonzero_ref.cl:24]: This is compile time preprocessor, not be able to be judged at runtime. I guess that your attention is to check this condition at runtime. For that purpose, USE_LOCAL_MEM should be handled as a variable, not the macro.

Comment (andrew-k-park) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/gather_nonzero_ref.cl]: updated as your suggestion.

**Review by yeonbok — APPROVED** [review]

**Pull request merged**